### ISSUE TITLE
Verify the expectations of the presentational role tests

### ISSUE THREAD

**Issue by joanmarie** [WhatSock/w3c-alternative-text-computation#10]
(Originally [posted by me on the public-aria list](https://lists.w3.org/Archives/Public/public-aria/2018Mar/0041.html))

I'm now working on the test cases where the expected result includes "Garaventa the weird." Both Firefox and WebKit have results which include "Eli" instead of "Garaventa." Looking at the AccName algorithm, I think Firefox and WebKit are correct.

Searching the algorithm, I see only one mention of the presentational role, namely in step 2D, which states:

> Otherwise, if the current node's native markup provides an attribute (e.g. title) or element (e.g. HTML label) that defines a text alternative, return that alternative in the form of a flat string as defined by the host language, unless the element is marked as presentational (role="presentation" or role="none").
> 
> Comment:
> For example, in HTML, the img element's alt attribute defines a text alternative string, and the label element provides text for the referenced form element. In SVG2, the desc and title elements provide a description of their parent element.

Looking at your test case, the markup relevant to which name is chosen is this bit:

```
<span role="presentation" aria-label="Eli">
  <span aria-label="Garaventa">Zambino</span>
</span>
```

While the presentational role is applied to the outer span, my understanding of an attribute provided by the node's native markup is something specific to HTML. The aria-label attribute can be used on HTML
of course; but the aria-label attribute is not *provided by* HTML. So my read of step 2D is that technically "Eli" should be chosen.

Furthermore, in the case where the thing being calculated is a name, I don't think we'd get to step 2D because step 2C states:

> Otherwise, if computing a name, and if the current node has an aria-label attribute whose value is not the empty string, nor, when trimmed of white space, is not the empty string:
> 
> * If traversal of the current node is due to recursion and the current node is an embedded control as defined in step 2E, ignore aria-label and skip to rule 2E.
> * Otherwise, return the value of aria-label.

Step 2C doesn't say anything about the presentational role. And the current node is not an embedded control as defined in step 2E. So why don't we want to use the aria-label value of "Eli"?

**Comment by accdc** [comment]
In reading this, and comparing it with the spec text, I still think my interpretation is correct.

For example, the relevant node that includes aria-label="Eli" is specifically this one:

```
<span role="presentation" aria-label="Eli">
```

So this is the only node that it applies to, because role=presentation or none only apply to that node and not to any of its children.

So if I change this to expose aria-label regardless of the presentational role, it will also expose the following node in the same way:

```
<img role="presentation" aria-label="Eli">
```

And the same for 

```
<img role="presentation" alt="Eli">
```

Is this what you expect?

**Comment by jnurthen** [comment]
As discussed in the meeting on 5 April, 2018, role=presentation should be ignored when a global aria state or property is applied. 
`<span role="presentation" aria-label="Eli">` 
would be the same as
`<span aria-label="Eli">`

`<img role="presentation" aria-label="Eli">`
would be the same as
`<img aria-label="Eli">`

and
`<img role="presentation" alt="Eli">`
would be the same as
`<img role="presentation" alt="Eli">`
as alt is not a global state or property so has no impact on role=presentation.

**Comment by joanmarie** [comment]
@jnurthen So to be clear, does that mean the expectations are currently incorrect and where "Garaventa" is currently, we should instead see "Eli". If that is indeed the case, once @accdc updates his tool, then I can update the wiki and WPT.

**Comment by jnurthen** [comment]
Correct. I believe this is waiting on @accdc to update the tool.

**Comment by joanmarie** [comment]
And the tool seems to be updated. Awesome and thanks @accdc!

I do have a follow-up question because while the problem described in the opening report of this issue seems fixed ("Eli" rather than "Garaventa"), there is a new issue which seems to be a side effect of the change made: the inclusion of "Bryan" in the resulting string, something I (and the implementations) disagree with.

The "Bryan" is coming from this element in the test cases:
```
<img src="file.jpg" title="Bryan" alt="" role="presentation" />
```

But here's [what the HTML-AAM says about name calculation for the `img` element](https://w3c.github.io/html-aam/#img-element):

> 1. If the `img` element has an `aria-label` or an `aria-labelledby` attribute the accessible name is to be calculated using the algorithm defined in Accessible Name and Description: Computation and API Mappings 1.1.
> 2. Otherwise use `alt` attribute
> 3. Otherwise use `title` attribute
> 4. If none of the above yield a usable text string there is no accessible name

1 doesn't apply. I would argue 2 _does_ apply: The author has explicitly set the `alt` attribute to the empty string. If there were no `alt` attribute, then I would fall through to 3 and use the `title` attribute. But with the explicit attribute having been provided, my interpretation of the HTML-AAM is that we have to use it and the final result would not contain "Bryan". Do you concur or do we need to solicit group input?

**Comment by jnurthen** [comment]
@joanmarie I agree with your assessment - @accdc what say you?

**Comment by accdc** [comment]
No problem, 2.1 should now have this fix. In running:

```
<img src="file.jpg" title="Bryan" alt="" role="presentation" id="test" />
```

The result is now:

accName: "Bryan"
accDesc: ""
(Running AccName Computation Prototype version: 2.1)


Changing which attributes have values will change the order of the computation as indicated in the list you pasted above, so this should now match. :)

**Comment by accdc** [comment]
Hi, I just pushed an update that I believe fixes the presentational issue regarding images. Basically, if role=presentation or role=none is present, it will be ignored as expected no matter if alt or title is present, however it will not be ignored if it includes a global ARIA attribute or if it is focusable. This conforms with all of the other ARIA criteria where presentational roles are applied.

**Comment by joanmarie** [comment]
Thanks @accdc . That looks more in keeping with what the user agents are doing. Is this clear from the AccName spec? And is it clear in the HTML-AAM?

**Comment by accdc** [comment]
In reading the spec, it looks clear to me.

I pushed an update this morning that simply streamlines the code for the Name and Description algorithms but there is no change to what is being computed.

If you see what you expect in this test, I believe this one is fixed. :)

**Comment by joanmarie** [comment]
OK, then closing this one as fixed. Thanks again for all your work and feedback on this @accdc!
